### ISSUE TITLE
coredump in `gms::gossiper::apply_new_states` while other node is draining

### ISSUE THREAD

**Issue by fruch** [scylladb/scylladb#21840]
## Packages

Scylla version: `6.3.0~dev-20241206.7e2875d6489d` with build-id `5227dd2a3fce4d2beb83ec6c17d47ad2e8ba6f5c`

Kernel Version: `6.8.0-1019-aws`

## Issue description

- [ ] This issue is a regression.
- [x] It is unknown if this issue is a regression.

While node-8 (10.4.21.244) was doing `nodetool drain`, node-6 (10.4.20.239) coredumped
with the following:

```
2024-12-09 01:46:05.391 <2024-12-09 01:46:05.288>: (DatabaseLogEvent Severity.ERROR) period_type=one-time event_id=81d79019-7739-4124-9db6-c7e58b7bcf0e: type=ABORTING_ON_SHARD regex=Aborting on shard line_number=22294 node=longevity-tls-50gb-3d-fix-mult-db-node-0362d13b-6
2024-12-09T01:46:05.288+00:00 longevity-tls-50gb-3d-fix-mult-db-node-0362d13b-6     !INFO | scylla[5632]: Aborting on shard 0, in scheduling group gossip.
void seastar::backtrace<seastar::backtrace_buffer::append_backtrace()::{lambda(seastar::frame)#1}>(seastar::backtrace_buffer::append_backtrace()::{lambda(seastar::frame)#1}&&) at reactor.cc:?
seastar::print_with_backtrace(seastar::backtrace_buffer&, bool) at reactor.cc:?
seastar::install_oneshot_signal_handler<6, (void (*)())(&seastar::sigabrt_action)>()::{lambda(int, siginfo_t*, void*)#1}::__invoke(int, siginfo_t*, void*) at reactor.cc:?
?? ??:0
?? ??:0
?? ??:0
?? ??:0
seastar::on_fatal_internal_error(seastar::logger&, std::basic_string_view<char, std::char_traits<char> >) at on_internal_error.cc:?
gms::gossiper::apply_new_states(gms::inet_address, gms::endpoint_state, gms::endpoint_state const&, utils::tagged_uuid<gms::permit_id_tag>, bool) [clone .resume] at gossiper.cc:?
seastar::internal::coroutine_traits_base<void>::promise_type::run_and_dispose() at main.cc:?
seastar::reactor::run_tasks(seastar::reactor::task_queue&) at reactor.cc:?
seastar::reactor::run_some_tasks() at reactor.cc:?
seastar::reactor::do_run() at reactor.cc:?
seastar::reactor::run() at reactor.cc:?
seastar::app_template::run_deprecated(int, char**, std::function<void ()>&&) at app-template.cc:?
seastar::app_template::run(int, char**, std::function<seastar::future<int> ()>&&) at app-template.cc:?
scylla_main(int, char**) at main.cc:?
std::function<int (int, char**)>::operator()(int, char**) const at ??:?
main at main.cc:?
?? ??:0
?? ??:0
_start at ??:?
``` 

```
2024-12-09T01:46:05.289+00:00 longevity-tls-50gb-3d-fix-mult-db-node-0362d13b-6      !ERR | scylla[5632]:  [shard  0: gms] gossip - Gossip change listener failed: std::bad_optional_access (bad optional access), at: 0x62dac8e 0x62db2b0 0x62db5b8 0x5d9a95e 0x5d9ab17 0x3cda720 0x145088a 0x5ddaddf 0x5ddc37a 0x5ddd557 0x5ddc908 0x5d6a583 0x5d698e3 0x13fc788 0x13fe180 0x13faa73 /opt/scylladb/libreloc/libc.so.6+0x2a087 /opt/scylladb/libreloc/libc.so.6+0x2a14a 0x13f8a24
   --------
   seastar::internal::coroutine_traits_base<void>::promise_type
   --------
   seastar::internal::coroutine_traits_base<void>::promise_type
   --------
   seastar::continuation<seastar::internal::promise_base_with_type<void>, seastar::future<void>::finally_body<seastar::with_semaphore<seastar::semaphore_default_exception_factory, gms::gossiper::apply_state_locally(std::map<gms::inet_address, gms::endpoint_state, std::less<gms::inet_address>, std::allocator<std::pair<gms::inet_address const, gms::endpoint_state> > >)::$_0::operator()<gms::inet_address&>(gms::inet_address&) const::{lambda()#1}, std::chrono::_V2::steady_clock>(seastar::basic_semaphore<seastar::semaphore_default_exception_factory, std::chrono::_V2::steady_clock>&, unsigned long, gms::gossiper::apply_state_locally(std::map<gms::inet_address, gms::endpoint_state, std::less<gms::inet_address>, std::allocator<std::pair<gms::inet_address const, gms::endpoint_state> > >)::$_0::operator()<gms::inet_address&>(gms::inet_address&) const::{lambda()#1}&&)::{lambda(auto:1)#1}::operator()<seastar::semaphore_units<seastar::semaphore_default_exception_factory, std::chrono::_V2::steady_clock> >(seastar::semaphore_units<seastar::semaphore_default_exception_factory, std::chrono::_V2::steady_clock>)::{lambda()#1}, false>, seastar::future<void>::then_wrapped_nrvo<seastar::future<void>, seastar::future<void>::finally_body<seastar::with_semaphore<seastar::semaphore_default_exception_factory, gms::gossiper::apply_state_locally(std::map<gms::inet_address, gms::endpoint_state, std::less<gms::inet_address>, std::allocator<std::pair<gms::inet_address const, gms::endpoint_state> > >)::$_0::operator()<gms::inet_address&>(gms::inet_address&) const::{lambda()#1}, std::chrono::_V2::steady_clock>(seastar::basic_semaphore<seastar::semaphore_default_exception_factory, std::chrono::_V2::steady_clock>&, unsigned long, gms::gossiper::apply_state_locally(std::map<gms::inet_address, gms::endpoint_state, std::less<gms::inet_address>, std::allocator<std::pair<gms::inet_address const, gms::endpoint_state> > >)::$_0::operator()<gms::inet_address&>(gms::inet_address&) const::{lambda()#1}&&)::{lambda(auto:1)#1}::operator()<seastar::semaphore_units<seastar::semaphore_default_exception_factory, std::chrono::_V2::steady_clock> >(seastar::semaphore_units<seastar::semaphore_default_exception_factory, std::chrono::_V2::steady_clock>)::{lambda()#1}, false> >(gms::gossiper::apply_state_locally(std::map<gms::inet_address, gms::endpoint_state, std::less<gms::inet_address>, std::allocator<std::pair<gms::inet_address const, gms::endpoint_state> > >)::$_0::operator()<gms::inet_address&>(gms::inet_address&) const::{lambda()#1}&&)::{lambda(seastar::internal::promise_base_with_type<void>&&, seastar::future<void>::finally_body<seastar::with_semaphore<seastar::semaphore_default_exception_factory, gms::gossiper::apply_state_locally(std::map<gms::inet_address, gms::endpoint_state, std::less<gms::inet_address>, std::allocator<std::pair<gms::inet_address const, gms::endpoint_state> > >)::$_0::operator()<gms::inet_address&>(auto:1&&) const::{lambda()#1}, std::chrono::_V2::steady_clock>(seastar::basic_semaphore<auto:1, auto:3>&, unsigned long, auto:2&&)::{lambda(auto:1)#1}::operator()<seastar::semaphore_units<seastar::semaphore_default_exception_factory, std::chrono::_V2::steady_clock> >(auto:1)::{lambda()#1}, false>&, seastar::future_state<seastar::internal::monostate>&&)#1}, void>
   --------
   seastar::coroutine::parallel_for_each<gms::gossiper::apply_state_locally(std::map<gms::inet_address, gms::endpoint_state, std::less<gms::inet_address>, std::allocator<std::pair<gms::inet_address const, gms::endpoint_state> > >)::$_0>
   --------
   seastar::internal::coroutine_traits_base<void>::promise_type
   --------
   seastar::internal::coroutine_traits_base<void>::promise_type
   --------
   N7seastar12continuationINS_8internal22promise_base_with_typeIvEENS_6futureIvE12finally_bodyIZNS_3smp9submit_toIZNS_7shardedIN3gms8gossiperEE9invoke_onINS_20noncopyable_functionIFS5_RSB_EEEJES5_Qsr3stdE9invocableITL0__RT_DpOTL0_0_EEET1_jNS_21smp_submit_to_optionsEOSJ_DpOT0_EUlvE_EENS_8futurizeINSt13invoke_resultISJ_JEE4typeEE4typeEjSP_SQ_EUlvE_Lb0EEEZNS5_17then_wrapped_nrvoIS5_S12_EENSV_ISJ_E4typeEOT0_EUlOS3_RS12_ONS_12future_stateINS1_9monostateEEEE_vEE
   --------
   N7seastar12continuationINS_8internal22promise_base_with_typeIvEEZNS_6futureIvE16handle_exceptionIZZN3gms8gossiper14background_msgENS_13basic_sstringIcjLj15ELb1EEENS_20noncopyable_functionIFS5_RS8_EEEEN3$_0clEvEUlT_E_Qoooooosr3stdE16is_invocable_r_vINS4_ISG_EETL0__NSt15__exception_ptr13exception_ptrEEaaeqsr3stdE12tuple_size_vINSt11conditionalIXsr3stdE9is_same_vINS1_18future_stored_typeIJSG_EE4typeENS1_9monostateEEESt5tupleIJEESS_IJSQ_EEE4typeEELi0Esr3stdE16is_invocable_r_vIvSK_SM_Eaaeqsr3stdE12tuple_size_vISW_ELi1Esr3stdE16is_invocable_r_vISG_SK_SM_Eaagtsr3stdE12tuple_size_vISW_ELi1Esr3stdE16is_invocable_r_vISW_SK_SM_EEES5_OSG_EUlSX_E_ZNS5_17then_wrapped_nrvoIS5_SY_EENS_8futurizeISG_E4typeEOT0_EUlOS3_RSY_ONS_12future_stateISR_EEE_vEE
   --------
   seastar::continuation<seastar::internal::promise_base_with_type<void>, seastar::future<void>::finally_body<seastar::internal::invoke_func_with_gate<gms::gossiper::background_msg(seastar::basic_sstring<char, unsigned int, 15u, true>, seastar::noncopyable_function<seastar::future<void> (gms::gossiper&)>)::$_0>(seastar::gate::holder&&, gms::gossiper::background_msg(seastar::basic_sstring<char, unsigned int, 15u, true>, seastar::noncopyable_function<seastar::future<void> (gms::gossiper&)>)::$_0&&)::{lambda()#1}, false>, seastar::future<void>::then_wrapped_nrvo<seastar::future<void>, seastar::future<void>::finally_body<seastar::internal::invoke_func_with_gate<gms::gossiper::background_msg(seastar::basic_sstring<char, unsigned int, 15u, true>, seastar::noncopyable_function<seastar::future<void> (gms::gossiper&)>)::$_0>(seastar::gate::holder&&, gms::gossiper::background_msg(seastar::basic_sstring<char, unsigned int, 15u, true>, seastar::noncopyable_function<seastar::future<void> (gms::gossiper&)>)::$_0&&)::{lambda()#1}, false> >(seastar::future<void>::finally_body<seastar::internal::invoke_func_with_gate<gms::gossiper::background_msg(seastar::basic_sstring<char, unsigned int, 15u, true>, seastar::noncopyable_function<seastar::future<void> (gms::gossiper&)>)::$_0>(seastar::gate::holder&&, gms::gossiper::background_msg(seastar::basic_sstring<char, unsigned int, 15u, true>, seastar::noncopyable_function<seastar::future<void> (gms::gossiper&)>)::$_0&&)::{lambda()#1}, false>&&)::{lambda(seastar::internal::promise_base_with_type<void>&&, seastar::future<void>::finally_body<seastar::internal::invoke_func_with_gate<gms::gossiper::background_msg(seastar::basic_sstring<char, unsigned int, 15u, true>, seastar::noncopyable_function<seastar::future<void> (gms::gossiper&)>)::$_0>(seastar::gate::holder&&, auto:1&&)::{lambda()#1}, false>&, seastar::future_state<seastar::internal::monostate>&&)#1}, void>
```

## Impact

lost two nodes at the same time, and quorum was lost

## How frequently does it reproduce?

first time we see this issue

## Installation details

Cluster size: 6 nodes (i4i.4xlarge)

Scylla Nodes used in this run:
- longevity-tls-50gb-3d-fix-mult-db-node-0362d13b-9 (54.170.35.129 | 10.4.23.10) (shards: 14)
- longevity-tls-50gb-3d-fix-mult-db-node-0362d13b-8 (52.19.153.205 | 10.4.21.244) (shards: 14)
- longevity-tls-50gb-3d-fix-mult-db-node-0362d13b-7 (34.240.36.206 | 10.4.21.114) (shards: -1)
- longevity-tls-50gb-3d-fix-mult-db-node-0362d13b-6 (34.249.43.57 | 10.4.20.239) (shards: 14)
- longevity-tls-50gb-3d-fix-mult-db-node-0362d13b-5 (54.77.175.32 | 10.4.22.111) (shards: 14)
- longevity-tls-50gb-3d-fix-mult-db-node-0362d13b-4 (52.213.100.43 | 10.4.21.2) (shards: 14)
- longevity-tls-50gb-3d-fix-mult-db-node-0362d13b-3 (52.16.233.231 | 10.4.22.114) (shards: 14)
- longevity-tls-50gb-3d-fix-mult-db-node-0362d13b-2 (34.243.24.234 | 10.4.20.129) (shards: 14)
- longevity-tls-50gb-3d-fix-mult-db-node-0362d13b-1 (54.73.67.181 | 10.4.20.132) (shards: 14)

OS / Image: `ami-0c7b4b0835c9342f7` (aws: undefined_region)

Test: `longevity-50gb-3days-test`
Test id: `0362d13b-3daa-436c-b416-edc58d9febc3`
Test name: `scylla-staging/fruch/longevity-50gb-3days-test`
Test method: `longevity_test.LongevityTest.test_custom_time`
Test config file(s):

- [longevity-50GB-3days-authorization-and-tls-ssl.yaml](https://github.com/scylladb/scylla-cluster-tests/blob/5eba5ae5da45fad05b5e2712ff714715f23563cb/test-cases/longevity/longevity-50GB-3days-authorization-and-tls-ssl.yaml)

<details>
<summary>
Logs and commands
</summary>

- Restore Monitor Stack command: `$ hydra investigate show-monitor 0362d13b-3daa-436c-b416-edc58d9febc3`
- Restore monitor on AWS instance using [Jenkins job](https://jenkins.scylladb.com/view/QA/job/QA-tools/job/hydra-show-monitor/parambuild/?test_id=0362d13b-3daa-436c-b416-edc58d9febc3)
- Show all stored logs command: `$ hydra investigate show-logs 0362d13b-3daa-436c-b416-edc58d9febc3`

## Logs:
- **longevity-tls-50gb-3d-fix-mult-db-node-0362d13b-7** - [https://cloudius-jenkins-test.s3.amazonaws.com/0362d13b-3daa-436c-b416-edc58d9febc3/20241208_202900/longevity-tls-50gb-3d-fix-mult-db-node-0362d13b-7-0362d13b.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/0362d13b-3daa-436c-b416-edc58d9febc3/20241208_202900/longevity-tls-50gb-3d-fix-mult-db-node-0362d13b-7-0362d13b.tar.gz)
- **longevity-tls-50gb-3d-fix-mult-db-node-0362d13b-1** - [https://cloudius-jenkins-test.s3.amazonaws.com/0362d13b-3daa-436c-b416-edc58d9febc3/20241208_202900/longevity-tls-50gb-3d-fix-mult-db-node-0362d13b-1-0362d13b.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/0362d13b-3daa-436c-b416-edc58d9febc3/20241208_202900/longevity-tls-50gb-3d-fix-mult-db-node-0362d13b-1-0362d13b.tar.gz)
- **db-cluster-0362d13b.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/0362d13b-3daa-436c-b416-edc58d9febc3/20241209_015856/db-cluster-0362d13b.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/0362d13b-3daa-436c-b416-edc58d9febc3/20241209_015856/db-cluster-0362d13b.tar.gz)
- **sct-runner-events-0362d13b.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/0362d13b-3daa-436c-b416-edc58d9febc3/20241209_015856/sct-runner-events-0362d13b.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/0362d13b-3daa-436c-b416-edc58d9febc3/20241209_015856/sct-runner-events-0362d13b.tar.gz)
- **sct-0362d13b.log.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/0362d13b-3daa-436c-b416-edc58d9febc3/20241209_015856/sct-0362d13b.log.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/0362d13b-3daa-436c-b416-edc58d9febc3/20241209_015856/sct-0362d13b.log.tar.gz)
- **loader-set-0362d13b.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/0362d13b-3daa-436c-b416-edc58d9febc3/20241209_015856/loader-set-0362d13b.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/0362d13b-3daa-436c-b416-edc58d9febc3/20241209_015856/loader-set-0362d13b.tar.gz)
- **monitor-set-0362d13b.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/0362d13b-3daa-436c-b416-edc58d9febc3/20241209_015856/monitor-set-0362d13b.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/0362d13b-3daa-436c-b416-edc58d9febc3/20241209_015856/monitor-set-0362d13b.tar.gz)
- **ssl-conf-0362d13b.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/0362d13b-3daa-436c-b416-edc58d9febc3/20241209_015856/ssl-conf-0362d13b.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/0362d13b-3daa-436c-b416-edc58d9febc3/20241209_015856/ssl-conf-0362d13b.tar.gz)

[Jenkins job URL](https://jenkins.scylladb.com/job/scylla-staging/job/fruch/job/longevity-50gb-3days-test/2/)
[Argus](https://argus.scylladb.com/test/18654ce0-fb23-4b05-8829-1aa321321294/runs?additionalRuns[]=0362d13b-3daa-436c-b416-edc58d9febc3)
</details>

**Comment by mykaul** [comment]
Added assignment and labels.

**Comment by kbr-scylla** [comment]
```cpp
    // Exceptions thrown from listeners will result in abort because that could leave the node in a bad
    // state indefinitely. Unless the value changes again, we wouldn't retry notifications.
    // Some values are set only once, so listeners would never be re-run.
    // Listeners should decide which failures are non-fatal and swallow them.
    try {
        co_await do_on_change_notifications(addr, changed, pid);
    } catch (...) {
        auto msg = format("Gossip change listener failed: {}", std::current_exception());
        if (_abort_source.abort_requested()) {
            logger.warn("{}. Ignored", msg);
        } else {
            on_fatal_internal_error(logger, msg);
        }
    }
```

so this had a `bad_optional_access`. We don't know where it's from because we don't have the backtrace of the original exception, it was swallowed, we only have the backtrace of the `on_fatal_internal_error` call which is useless.

**Comment by kbr-scylla** [comment]
@gleb-cloudius please triage this

**Comment by gleb-cloudius** [comment]
The crash is not the only place where we hit bad_optional_access:

```
longevity-tls-50gb-3d-fix-mult-db-node-0362d13b-2/messages.log:2024-12-08T23:48:41.269+00:00 longevity-tls-50gb-3d-fix-mult-db-node-0362d13b-2  !WARNING | scylla[2962]:  [shard  0: gms] gossip - Failed to handle GOSSIP_SHUTDOWN: std::bad_optional_access (bad optional access)
longevity-tls-50gb-3d-fix-mult-db-node-0362d13b-2/messages.log:2024-12-08T23:49:49.955+00:00 longevity-tls-50gb-3d-fix-mult-db-node-0362d13b-2  !WARNING | scylla[2962]:  [shard  0: gms] gossip - Failed to handle GOSSIP_DIGEST_ACK2: std::bad_optional_access (bad optional access)
longevity-tls-50gb-3d-fix-mult-db-node-0362d13b-2/messages.log:2024-12-09T01:04:50.013+00:00 longevity-tls-50gb-3d-fix-mult-db-node-0362d13b-2  !WARNING | scylla[2962]:  [shard  0: gms] gossip - Failed to handle GOSSIP_SHUTDOWN: std::bad_optional_access (bad optional access)
longevity-tls-50gb-3d-fix-mult-db-node-0362d13b-2/messages.log:2024-12-09T01:04:55.513+00:00 longevity-tls-50gb-3d-fix-mult-db-node-0362d13b-2  !WARNING | scylla[2962]:  [shard  0: gms] gossip - Failed to handle GOSSIP_DIGEST_ACK2: std::bad_optional_access (bad optional access)
longevity-tls-50gb-3d-fix-mult-db-node-0362d13b-6/messages.log:2024-12-08T23:48:41.289+00:00 longevity-tls-50gb-3d-fix-mult-db-node-0362d13b-6  !WARNING | scylla[5632]:  [shard  0: gms] gossip - Failed to handle GOSSIP_SHUTDOWN: std::bad_optional_access (bad optional access)
longevity-tls-50gb-3d-fix-mult-db-node-0362d13b-6/messages.log:2024-12-08T23:49:50.038+00:00 longevity-tls-50gb-3d-fix-mult-db-node-0362d13b-6  !WARNING | scylla[5632]:  [shard  0: gms] gossip - Failed to handle GOSSIP_DIGEST_ACK2: std::bad_optional_access (bad optional access)
longevity-tls-50gb-3d-fix-mult-db-node-0362d13b-6/messages.log:2024-12-09T01:04:50.039+00:00 longevity-tls-50gb-3d-fix-mult-db-node-0362d13b-6  !WARNING | scylla[5632]:  [shard  0: gms] gossip - Failed to handle GOSSIP_SHUTDOWN: std::bad_optional_access (bad optional access)
longevity-tls-50gb-3d-fix-mult-db-node-0362d13b-6/messages.log:2024-12-09T01:04:55.538+00:00 longevity-tls-50gb-3d-fix-mult-db-node-0362d13b-6  !WARNING | scylla[5632]:  [shard  0: gms] gossip - Failed to handle GOSSIP_DIGEST_ACK2: std::bad_optional_access (bad optional access)
longevity-tls-50gb-3d-fix-mult-db-node-0362d13b-6/messages.log:2024-12-09T01:36:00.789+00:00 longevity-tls-50gb-3d-fix-mult-db-node-0362d13b-6  !WARNING | scylla[5632]:  [shard  0: gms] gossip - Failed to handle GOSSIP_SHUTDOWN: std::bad_optional_access (bad optional access)
longevity-tls-50gb-3d-fix-mult-db-node-0362d13b-6/messages.log:2024-12-09T01:37:11.789+00:00 longevity-tls-50gb-3d-fix-mult-db-node-0362d13b-6  !WARNING | scylla[5632]:  [shard  0: gms] gossip - Failed to handle GOSSIP_DIGEST_ACK: std::bad_optional_access (bad optional access)
longevity-tls-50gb-3d-fix-mult-db-node-0362d13b-6/messages.log:2024-12-09T01:46:05.289+00:00 longevity-tls-50gb-3d-fix-mult-db-node-0362d13b-6      !ERR | scylla[5632]:  [shard  0: gms] gossip - Gossip change listener failed: std::bad_optional_access (bad optional access), at: 0x62dac8e 0x62db2b0 0x62db5b8 0x5d9a95e 0x5d9ab17 0x3cda720 0x145088a 0x5ddaddf 0x5ddc37a 0x5ddd557 0x5ddc908 0x5d6a583 0x5d698e3 0x13fc788 0x13fe180 0x13faa73 /opt/scylladb/libreloc/libc.so.6+0x2a087 /opt/scylladb/libreloc/libc.s
```

What is common between all of those gossiper messages is that we may call `messaging_service::remove_rpc_client(ip)`  while handling them. 

The function now has a loop over clients that were opened using host id as dst address. To find which of those have provided IP there is a call to `_address_map.find(hid).value()` that assumes that mapping exists. It should because we successfully opened a connection to this hid and it means that during connection creation the mapping was available. If host is removed all the connection to it should be dropped.

So if this is the path that throws it is possible that we have a race where we can open a connection to a node that is being decommissioned after we already dropped all the connection to it but before it was actually removed. Trying to map host id to ip for such connections will eventually start to fail.
